CTFd/api/v1/challenges.py

```diff
@@ -75,6 +76,43 @@ class ChallengeListSuccessResponse(APIListSuccessResponse):
 )
 
 
+def _build_solves_query(extra_filters=tuple(), admin_view=False):
+    # This can return None (unauth) if visibility is set to public
+    user = get_current_user()
+    # We only set a condition for matching user solves if there is a user and
+    # they have an account ID (user mode or in a team in teams mode)
+    if user is not None and user.account_id is not None:
+        user_solved_cond = (Solves.account_id == user.account_id)
+    else:
+        user_solved_cond = false()
+    # We have to filter solves to exclude any made after the current freeze
+    # time unless we're in an admin view as determined by the caller.
+    freeze = get_config("freeze")
+    if freeze and not admin_view:
+        freeze_cond = (Solves.date < unix_time_to_utc(freeze))
+    else:
+        freeze_cond = true()
+    # Finally, we never count solves made by hidden or banned users/teams, even
+    # if we are an admin. This is to match the challenge detail API.
+    AccountModel = get_model()
+    exclude_solves_cond = and_(
+        AccountModel.banned == false(),
+        AccountModel.hidden == false(),
+    )
+    # This query counts the number of solves per challenge, as well as the sum
+    # of correct solves made by the current user per the condition above (which
+    # should probably only be 0 or 1!)
+    solves_q = db.session.query(
+        Solves.challenge_id, sa_func.count(Solves.challenge_id),
+        sa_func.sum(user_solved_cond),
+    ).join(AccountModel).filter(
```

Comment:
Note that I don't bother joining the challenges table to check for hidden challenges. I'm assuming that the caller will take care of not using data this query returns for challenges which are hidden or don't meet other visibility criteria.